### PR TITLE
Add mortar periodicity action

### LINKED ISSUE #7419: Add action for Mortar based periodicity
### Description of the enhancement or error report
Mortar/Constraint based periodicity like `EqualGradientConstraint` need to be significantly simpler to use. I propose adding an action to set up everything.

### Rationale for the enhancement or information for reproducing the error
Ease of use.

### Identified impact
Simplify input files

### PR DESCRIPTION
**Pull request by dschwen**
This should make it much easier to add the mortar based gradient periodicity.

Closes #7419

### CODE REVIEW

**Comment by dschwen** [comment]
I'll add a commit with the docs in a minute

**Comment by mangerij** [comment]
Great!

File: modules/phase_field/src/action/MortarPeriodicAction.C [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+      if (_current_task == "add_constraint")
+      {
+        for (auto & var : _variables)
+        {
+          const std::string ct_base = "ct_" + periodicity_name + "_" + boundary_names[i] + "_" + var;
+          switch (_periodicity)
+          {
+            case 0: // gradient
+              for (unsigned int j = 0; j < dim; ++j)
```
Comment (permcody) [inline modules/phase_field/src/action/MortarPeriodicAction.C]: Nobody gets types correct (unsigned int here).

File: modules/phase_field/src/action/MortarPeriodicAction.C [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  // boundaries
+  const std::vector<BoundaryName> boundary_names = { "left", "bottom", "back" };
+  // opposite boundaries
+  const std::vector<BoundaryName> opposite_boundary_names = { "right", "top", "front" };
+
+  // mortar subdomains
+  const std::vector<SubdomainID> & mortar_subdomains = mesh->getMortarSubdomains();
+
+  // iterate over the periodic directions
+  for (unsigned short i = 0; i < dim; ++i)
```
Comment (permcody) [inline modules/phase_field/src/action/MortarPeriodicAction.C]: unsigned short here

File: modules/phase_field/include/action/MortarPeriodicAction.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ */
+class MortarPeriodicAction : public Action
+{
+public:
+  MortarPeriodicAction(const InputParameters & parameters);
+
+  virtual void act();
+
+  // all variables this action operates on
+  std::vector<VariableName> _variables;
```
Comment (permcody) [inline modules/phase_field/include/action/MortarPeriodicAction.h]: Why is this public?

File: modules/phase_field/include/action/MortarPeriodicAction.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+  // type of the periodic constraint to apply (value, gradient)
+  const unsigned int _periodicity;
```
Comment (permcody) [inline modules/phase_field/include/action/MortarPeriodicAction.h]: and this public?

File: modules/phase_field/src/action/MortarPeriodicAction.C (hunk 1 above)
Comment (dschwen) [inline modules/phase_field/src/action/MortarPeriodicAction.C]: Switched them all to `unsigned int` (which is what `dimension()` returns)

File: modules/phase_field/include/action/MortarPeriodicAction.h (hunk 4 above)
Comment (dschwen) [inline modules/phase_field/include/action/MortarPeriodicAction.h]: fixed

File: modules/phase_field/src/action/MortarPeriodicAction.C (hunk 1 above)
Comment (dschwen) [inline modules/phase_field/src/action/MortarPeriodicAction.C]: > Nobody gets types correct

Grrr, `MultiMooseEnum::contains()` accepts arguments of type `int` or `unsigned short`. It fails to compile when passing in an `unsigned int`...

File: modules/phase_field/include/action/MortarPeriodicAction.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * Set up Mortar based periodicity in an input file with a MortarPeriodicMesh
+ */
+class MortarPeriodicAction : public Action
+{
+public:
+  MortarPeriodicAction(const InputParameters & parameters);
+
+  virtual void act();
```
Comment (permcody) [inline modules/phase_field/include/action/MortarPeriodicAction.h]: override!

**Comment by permcody** [comment]
👍

**Pull request merged**